src/lustre/lustreDesugarFrameBlocks.ml

```diff
@@ -0,0 +1,264 @@
+(* This file is part of the Kind 2 model checker.
+
+   Copyright (c) 2022 by the Board of Trustees of the University of Iowa
+
+   Licensed under the Apache License, Version 2.0 (the "License"); you
+   may not use this file except in compliance with the License.  You
+   may obtain a copy of the License at
+
+   http://www.apache.org/licenses/LICENSE-2.0 
+
+   Unless required by applicable law or agreed to in writing, software
+   distributed under the License is distributed on an "AS IS" BASIS,
+   WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or
+   implied. See the License for the specific language governing
+   permissions and limitations under the License. 
+ *)
+
+module A = LustreAst
+module R = Res
+module LDI = LustreDesugarIfBlocks
+module Chk = LustreTypeChecker
+
+let (let*) = R.(>>=)
+
+type error_kind = 
+  | MisplacedNodeItemError of A.node_item
+  | MisplacedFrameBlockError of A.node_item
+
+let error_message error = match error with
+  | MisplacedNodeItemError ni -> (match ni with
+    | Body (Assert _) -> "Asserts are not allowed inside frame blocks."
+    | FrameBlock _ -> "Frame blocks are not allowed inside other frame blocks."
+    | AnnotMain _ -> "Main annotations are not allowed inside frame blocks."
+    | AnnotProperty _ -> "Property annotations are not allowed inside frame blocks."
+    (* Other node items are allowed *)
+    | _ -> assert false
+    )
+  | MisplacedFrameBlockError _ -> "Frame blocks are not allowed in functions."
+
+type error = [
+  | `LustreDesugarFrameBlocksError of Lib.position * error_kind
+  | `LustreDesugarIfBlocksError of Lib.position * LDI.error_kind
+  | `LustreAstInlineConstantsError of Lib.position * LustreAstInlineConstants.error_kind
+  | `LustreSyntaxChecksError of Lib.position * LustreSyntaxChecks.error_kind
+  | `LustreTypeCheckerError of Lib.position * LustreTypeChecker.error_kind
+]
+
+let mk_error pos kind = Error (`LustreDesugarFrameBlocksError (pos, kind))
+
+(** Parses an expression and replaces any ITE oracles with the 'fill'
+    expression (which is stuttering, ie, 'pre variable').
+*)
+let rec fill_ite_helper fill = function
+  (* Replace all oracles with 'fill' *)
+  | A.Ident (pos, i) -> 
+    (* See if 'i' is of the form "n_iboracle" *)
+    let str = String.split_on_char '_' (HString.string_of_hstring i) |>
+      List.rev |> List.hd
+    in
+    if (str = "iboracle") then fill else A.Ident(pos, i)
```

Comment:
The check `str = "iboracle"` creates a logic dependency between modules (`LustreDesugarFrameBlocks` and `LustreDesugarIfBlocks`). You should put the definition of the constant `"iboracle"` and this check in a common place, outside both modules, so that the dependency is clear. Perhaps `GeneratedIdentifiers` is a good place. If not, choose a different module which makes sense or create a new module.